File: diagnostics/j2534-service/src/J2534.Host/CapabilityToken.cs
using System.Collections.Concurrent;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace MechPro.J2534.Host;

/// <summary>
/// HMAC-signed clear_dtcs capability tokens. Compatible with
/// infra/lambda/diagnostics/capability-token.ts and the Node host verifier.
/// </summary>
public static class CapabilityToken
{
    static readonly ConcurrentDictionary<string, byte> Consumed = new();

    public static void VerifyClearDtcs(string? token)
    {
        var raw = (token ?? string.Empty).Trim();
        var parts = raw.Split('.');
        if (parts.Length != 3 || parts[0] != "v1")
        {
            throw new UnauthorizedAccessException("Invalid diagnostics capability token");
        }

        var payloadJson = Encoding.UTF8.GetString(Base64UrlDecode(parts[1]));
        var expectedSig = Sign(payloadJson);
        if (!FixedTimeEquals(expectedSig, parts[2]))
        {
            throw new UnauthorizedAccessException("Invalid diagnostics capability token signature");
        }

        using var doc = JsonDocument.Parse(payloadJson);
        var root = doc.RootElement;
        if (root.GetProperty("v").GetInt32() != 1
            || root.GetProperty("procedure").GetString() != "clear_dtcs")
        {
            throw new UnauthorizedAccessException("Capability token is not valid for clearDtcs");
        }

        var vin = root.GetProperty("vin").GetString();
        var shopId = root.GetProperty("shopId").GetString();
        var jti = root.GetProperty("jti").GetString();
        var exp = root.GetProperty("exp").GetInt64();
        if (string.IsNullOrWhiteSpace(vin) || string.IsNullOrWhiteSpace(shopId) || string.IsNullOrWhiteSpace(jti))
        {
            throw new UnauthorizedAccessException("Capability token payload is incomplete");
        }

        if (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() > exp)
        {
            throw new UnauthorizedAccessException("Capability token expired");
        }

        if (!Consumed.TryAdd(jti, 0))
        {
            throw new UnauthorizedAccessException("Capability token already used");
        }

        if (Consumed.Count > 500)
        {
            foreach (var key in Consumed.Keys.Take(50))
            {
                Consumed.TryRemove(key, out _);
            }
        }
    }

    static string Sign(string payloadJson)
    {
        var secret = Environment.GetEnvironmentVariable("MECHPRO_DIAG_CAPABILITY_SECRET")
            ?? Environment.GetEnvironmentVariable("DIAGNOSTICS_CAPABILITY_SECRET")
            ?? "mechpro-dev-diagnostics-capability-v1";
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
        return Base64UrlEncode(hmac.ComputeHash(Encoding.UTF8.GetBytes(payloadJson)));
    }

    static bool FixedTimeEquals(string a, string b)
    {
        var left = Encoding.UTF8.GetBytes(a);
        var right = Encoding.UTF8.GetBytes(b);
        return left.Length == right.Length && CryptographicOperations.FixedTimeEquals(left, right);
    }

    static string Base64UrlEncode(byte[] data) =>
        Convert.ToBase64String(data).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    static byte[] Base64UrlDecode(string input)
    {
        var padded = input.Replace('-', '+').Replace('_', '/');
        switch (padded.Length % 4)
        {
            case 2: padded += "=="; break;
            case 3: padded += "="; break;
        }
        return Convert.FromBase64String(padded);
    }
}

File: diagnostics/j2534-host-node/capability-token.js
/**
 * HMAC-signed diagnostics capability tokens (clear_dtcs).
 * Must stay compatible with infra/lambda/diagnostics/capability-token.ts
 */
const { createHmac, randomBytes, timingSafeEqual } = require('node:crypto');

const consumedJti = new Set();

function secret() {
  return String(
    process.env.MECHPRO_DIAG_CAPABILITY_SECRET
    || process.env.DIAGNOSTICS_CAPABILITY_SECRET
    || 'mechpro-dev-diagnostics-capability-v1',
  ).trim();
}

function signPayload(payloadJson) {
  return createHmac('sha256', secret()).update(payloadJson).digest('base64url');
}

function verifyClearDtcsToken(token, expected = {}) {
  const raw = String(token || '').trim();
  const parts = raw.split('.');
  if (parts.length !== 3 || parts[0] !== 'v1') {
    throw new Error('Invalid diagnostics capability token');
  }
  const payloadJson = Buffer.from(parts[1], 'base64url').toString('utf8');
  const expectedSig = signPayload(payloadJson);
  const providedSig = parts[2];
  const expectedBuf = Buffer.from(expectedSig, 'base64url');
  const providedBuf = Buffer.from(providedSig, 'base64url');
  if (expectedBuf.length !== providedBuf.length || !timingSafeEqual(expectedBuf, providedBuf)) {
    throw new Error('Invalid diagnostics capability token signature');
  }
  const payload = JSON.parse(payloadJson);
  if (payload.v !== 1 || payload.procedure !== 'clear_dtcs') {
    throw new Error('Capability token is not valid for clearDtcs');
  }
  if (!payload.vin || !payload.shopId || !payload.jti || !payload.exp) {
    throw new Error('Capability token payload is incomplete');
  }
  if (Date.now() > Number(payload.exp)) {
    throw new Error('Capability token expired');
  }
  if (expected.vin && payload.vin !== String(expected.vin).trim().toUpperCase()) {
    throw new Error('Capability token VIN mismatch');
  }
  if (expected.consume !== false) {
    if (consumedJti.has(payload.jti)) {
      throw new Error('Capability token already used');
    }
    consumedJti.add(payload.jti);
    if (consumedJti.size > 500) {
      const first = consumedJti.values().next().value;
      consumedJti.delete(first);
    }
  }
  return payload;
}

function mintClearDtcsToken(input, ttlMs = 5 * 60 * 1000) {
  const payload = {
    v: 1,
    procedure: 'clear_dtcs',
    vin: String(input.vin || '').trim().toUpperCase(),
    shopId: String(input.shopId || 'local').trim(),
    exp: Date.now() + ttlMs,
    jti: randomBytes(12).toString('hex'),
  };
  const payloadJson = JSON.stringify(payload);
  const token = `v1.${Buffer.from(payloadJson).toString('base64url')}.${signPayload(payloadJson)}`;
  return { token, expiresAt: new Date(payload.exp).toISOString(), payload };
}

module.exports = { verifyClearDtcsToken, mintClearDtcsToken };
